dagshub/auth/tokens.py

```diff
@@ -0,0 +1,126 @@
+        self.cache_location = cache_location
+        self.__token_cache: Optional[Dict[str, List[Dict]]] = None
+
+    @property
+    def _token_cache(self):
+        if self.__token_cache is None:
+            self.__token_cache = self._load_cache_file()
+        return self.__token_cache
+
+    def add_token(self, token: Dict, host: str = None):
+        if host is None:
+            host = config.host
+        if host not in self._token_cache:
+            self._token_cache[host] = []
+        self._token_cache[host].append(token)
+        self._store_cache_file()
+
+    def get_token(self, host: str = None):
+        if host is None:
+            host = config.host
+        tokens = self._token_cache.get(host, [])
+        app_tokens = list(filter(lambda t: t["token_type"] == APP_TOKEN_TYPE, tokens))
+        if len(app_tokens) > 0:
+            token = app_tokens[0]
+        else:
+            non_expired_tokens = list(filter(lambda t: not self._is_expired(t), tokens))
+            if len(non_expired_tokens) > 0:
+                token = non_expired_tokens[0]
+            else:
+                logger.warning(f"No valid tokens left for host '{host}'. Reauthenticating with OAuth")
+                token = oauth.oauth_flow(host)
+                tokens.append(token)
+                self._token_cache[host] = tokens
+                self._store_cache_file()
+        return token["access_token"]
+
+    @staticmethod
+    def _is_expired(token: Dict[str, str]) -> bool:
+        if "expiry" not in token:
+            return True
+        if token["expiry"] == "never":
+            return False
+        # Need to cut off the three additional precision numbers in milliseconds, because %f only parses 6 digits
+        expiry = token["expiry"][:-4] + "Z"
+        expiry_dt = datetime.datetime.strptime(expiry, "%Y-%m-%dT%H:%M:%S.%fZ")
+        is_expired = expiry_dt < datetime.datetime.utcnow()
+        return is_expired
+
+    def _load_cache_file(self) -> Dict[str, List[Dict]]:
+        logger.debug(f"Loading OAuth token cache from {self.cache_location}")
+        if not os.path.exists(self.cache_location):
+            logger.debug("OAuth token cache file doesn't exist")
+            return {"version": config.CONFIG_SCHEMA_VERSION}
+        try:
+            with open(self.cache_location) as f:
+                tokens_cache = yaml.load(f, yaml.Loader)
+                return tokens_cache
+        except:
+            logger.error(
+                f"Error while loading DagsHub OAuth token cache: {traceback.format_exc()}"
+            )
+            raise
+
+    def _store_cache_file(self):
+        logger.debug(f"Dumping OAuth token cache to {self.cache_location}")
+        try:
+            dirpath = os.path.dirname(self.cache_location)
+            if not os.path.exists(dirpath):
+                os.makedirs(dirpath)
+            with open(self.cache_location, "w") as f:
+                yaml.dump(self.__token_cache, f, yaml.Dumper)
```

Comment:
Not critical now I guess, but this loading and saving flow is not safe for parallel work on the same cache file. The latest to save will override whatever is in there.

Reply:
What's the solution? Re-cache the file every time before saving and adding the difference?

Reply:
I don't really think a solution is necessary rn, if that become a problem we'll think of one...